### PR TITLE
feat: add G.711 a-law algorithm

### PR DESCRIPTION
**Pull request by HiSunzhenliang**
#### Description of Change
1. add G.711 a-law algorithm
2. add audio/

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C/pull/858"></a>

### CODE REVIEW

**Comment by HiSunzhenliang** [comment]
Please review this and let me know if it needs to be fixed @Panquesito7
Thanks.

File: audio/alaw.c [hunk 1]
```diff
@@ -0,0 +1,126 @@
+/**
+ * @file
+ * @author [sunzhenliang](https://github.com/HiSunzhenliang)
+ * @brief A-law algorithm for encoding and decoding (16bit pcm <=> a-law).
+ * @see audio/alaw.c
```
Comment (Panquesito7) [inline audio/alaw.c:5]: I think this is used when there are two or more similar files, so I don't think this is needed here. 🙂
```suggestion
```

File: audio/alaw.c [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+ * s0000000abcdx     | s000abcd        | s0000000abcd1
+ * s0000001abcdx     | s001abcd        | s0000001abcd1
+ * s000001abcdxx     | s010abcd        | s000001abcd10
+ * s00001abcdxxx     | s011abcd        | s00001abcd100
+ * s0001abcdxxxx     | s100abcd        | s0001abcd1000
+ * s001abcdxxxxx     | s101abcd        | s001abcd10000
+ * s01abcdxxxxxx     | s110abcd        | s01abcd100000
+ * s1abcdxxxxxxx     | s111abcd        | s1abcd1000000
+ *
+ * https://en.wikipedia.org/wiki/G.711
```
Comment (Panquesito7) [inline audio/alaw.c:20]:
```suggestion
```

File: audio/alaw.c [hunk 3]
```diff
@@ -0,0 +1,126 @@
+/**
+ * @file
+ * @author [sunzhenliang](https://github.com/HiSunzhenliang)
+ * @brief A-law algorithm for encoding and decoding (16bit pcm <=> a-law).
```
Comment (Panquesito7) [inline audio/alaw.c:4]:
```suggestion
 * @brief A-law algorithm for encoding and decoding (16bit pcm <=> a-law).
 * This is the implementation of [G.711](https://en.wikipedia.org/wiki/G.711) in C++
```

File: audio/alaw.c [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+ * s001abcdxxxxx     | s101abcd        | s001abcd10000
+ * s01abcdxxxxxx     | s110abcd        | s01abcd100000
+ * s1abcdxxxxxxx     | s111abcd        | s1abcd1000000
+ *
+ * https://en.wikipedia.org/wiki/G.711
+ *
+ * Compressed code: (s | eee | abcd)
+ **/
+#include <stdio.h>
+#include <stdlib.h>
```
Comment (Panquesito7) [inline audio/alaw.c:25]: Please add a one-line description of what the library/header is for (**see the example below**).

```c
#include <stdlib.h>     /// for IO operations
#include <assert.h>    /// for assert
```

File: audio/alaw.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        pcm += 0x8;
+        pcm += eee ? 0x100 : 0x0;
+        pcm <<= eee > 1 ? (eee - 1) : 0;
+
+        *out++ = sign ? -pcm : pcm;
+    }
+}
+
+#define LEN 4
+int main(int argc, char *argv[])
```
Comment (Panquesito7) [inline audio/alaw.c:96]:
```suggestion
/**
 * @brief Main function
 * @param argc commandline argument count (ignored)
 * @param argv commandline array of arguments (ignored)
 * @returns 0 on exit
 */
int main(int argc, char *argv[])
```

File: audio/alaw.c [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+{
+    short pcm[LEN] = {1000, 1234, 3200, -1314};
+    unsigned char coded[LEN];
+    short decoded[LEN];
+
+    printf("inputs: ");
+    for (int i = 0; i < LEN; i++)
```
Comment (Panquesito7) [inline audio/alaw.c:103]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). Requires adding the `inttypes.h` header. **Let us know if you need any help in our [Discord server](https://discord.gg/c7MnfGFGa6)**. 🙂

File: audio/alaw.c [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        sign = alaw & 0x80;
+        eee = (alaw & 0x70) >> 4;
+        pcm = alaw & 0x0f;
+        pcm <<= 4;
+        pcm += 0x8;
+        pcm += eee ? 0x100 : 0x0;
+        pcm <<= eee > 1 ? (eee - 1) : 0;
+
+        *out++ = sign ? -pcm : pcm;
+    }
```
Comment (Panquesito7) [inline audio/alaw.c:92]: Please add a bit of documentation of what this code does here.

File: audio/alaw.c [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    printf("\n");
+
+    printf("decode: ");
+    decode(decoded, coded, LEN);
+    for (int i = 0; i < LEN; i++)
+    {
+        printf("%d ", decoded[i]);
+    }
+    printf("\n");
```
Comment (Panquesito7) [inline audio/alaw.c:123]: Please add a bit of documentation of what this code does here.

File: audio/alaw.c [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        while ((pcm & mask) == 0 && eee > 0)
+        {
+            eee--;
+            mask >>= 1;
+        }
+        abcd = (pcm >> (eee ? (eee + 3) : 4)) & 0x0f;
+        eee <<= 4;
+
+        alaw = (sign | eee | abcd);
+        *out++ = alaw ^ 0xD5;
```
Comment (Panquesito7) [inline audio/alaw.c:62]: Please add a bit of documentation of what this code does here.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Woah, awesome work! 👀🚀

**Comment by HiSunzhenliang** [comment]
Thanks for reviewing. I made some changes and submitted it again.

File: audio/alaw.c [hunk 10]
```diff
@@ -0,0 +1,180 @@
+/**
+ * @file
+ * @author [sunzhenliang](https://github.com/HiSunzhenliang)
+ * @brief A-law algorithm for encoding and decoding (16bit pcm <=> a-law).
```
[the remark this answers is not in the document]
Comment (Panquesito7) [reply audio/alaw.c:4]: > Sorry, I can not see the problem here. Can you be more specific？Thanks.

Check line 5 in the suggestion. I hope that helps. 🙂

[the remark this answers is not in the document]
Comment (HiSunzhenliang) [reply audio/alaw.c:4]: I already added `This is the implementation of [G.711](https://en.wikipedia.org/wiki/G.711) in C.` at line 5. The only difference is `...in C++` in the suggestion from my comment `...in C`. And this is C code right? So I am quite confused. :confounded:

[the remark this answers is not in the document]
Comment (Panquesito7) [reply audio/alaw.c:4]: Ahh, sorry! I modified the suggestion. Great catch! 😅

File: audio/alaw.c [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+        pcm <<= eee > 1 ? (eee - 1) : 0;
+
+        /* Use the right sign */
+        *out++ = sign ? -pcm : pcm;
+    }
+}
+
+/* length of test inputs */
+#define LEN ((size_t)8)
+
```
Comment (Panquesito7) [inline audio/alaw.c:129]: Please add self-test implementations of your code to ensure it runs properly (**see the [typical structure of a program](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) for how to make it properly).

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: audio/alaw.c [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+ * @returns void
+ */
+void encode(uint8_t *out, int16_t *in, size_t len)
+{
+    uint8_t alaw;
+    int16_t pcm;
+    int32_t sign;
+    int32_t abcd;
+    int32_t eee;
+    int32_t mask;
```
Comment (Panquesito7) [inline audio/alaw.c:53]: You may want to initialize these variables to zero.
```suggestion
    uint8_t alaw = 0;
    int16_t pcm = 0;
    int32_t sign = 0;
    int32_t abcd = 0;
    int32_t eee = 0;
    int32_t mask = 0;
```

File: audio/alaw.c [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+ * @param in  unsigned 8bit alaw array
+ * @param len length of alaw array
+ * @returns void
+ */
+void decode(int16_t *out, uint8_t *in, size_t len)
+{
+    uint8_t alaw;
+    int32_t pcm;
+    int32_t sign;
+    int32_t eee;
```
Comment (Panquesito7) [inline audio/alaw.c:112]: Same here.
```suggestion
    uint8_t alaw = 0;
    int32_t pcm = 0;
    int32_t sign = 0;
    int32_t eee = 0;
```

File: audio/alaw.c [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+int main(int argc, char *argv[])
+{
+    /* output alaw encoded by encode() */
+    uint8_t coded[LEN];
+
+    /* output pcm decoded by decode() from coded[LEN] */
+    int16_t decoded[LEN];
+
+    /* run self-test implementations */
+    test(pcm, coded, decoded, LEN);
```
Comment (Panquesito7) [inline audio/alaw.c:184]:
```suggestion
    test(pcm, coded, decoded, LEN);  // run self-test implementations
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there; great work! 🚀

**Review by Panquesito7 — APPROVED** [review]
Awesome work! It's been great for your first contribution here! We hope you keep contributing. Thanks! 😄👍🎉

**Review by amino19 — APPROVED** [review]

**Review by mishraabhinn — APPROVED** [review]

**Review by ayaankhan98 — APPROVED** [review]
LGTM :+1:

**Comment by Panquesito7** [comment]
Thanks for reviewing, folks! 😄

**Pull request merged**